linkerd/app/src/lib.rs

```diff
@@ -146,30 +152,55 @@ impl Config {
-                    outbound::server::stack(
-                        &outbound,
-                        dst.profiles.clone(),
-                        dst.resolve,
-                        outbound::tcp::connect::stack(
-                            &outbound.proxy.connect,
-                            outbound_addr.port(),
-                            local_identity.clone(),
+            info!(listen.addr = %outbound_addr, ingress_mode);
+            if ingress_mode {
+                tokio::spawn(
+                    serve::serve(
+                        outbound_listen,
+                        outbound::ingress::stack(
+                            &outbound,
+                            dst.profiles.clone(),
+                            outbound::tcp::connect::forward(outbound::tcp::connect::stack(
+                                &outbound.proxy.connect,
+                                outbound_addr.port(),
+                                local_identity.clone(),
+                                &outbound_metrics,
+                            )),
+                            outbound_http.clone(),
                             &outbound_metrics,
+                            oc_span_sink.clone(),
+                            drain_rx.clone(),
                         ),
-                        outbound_http.clone(),
-                        outbound_metrics,
-                        oc_span_sink.clone(),
-                        drain_rx.clone(),
-                    ),
-                    drain_rx.clone().signal(),
-                )
-                .map_err(|e| panic!("outbound failed: {}", e))
-                .instrument(span.clone()),
-            );
+                        drain_rx.clone().signal(),
+                    )
+                    .map_err(|e| panic!("outbound failed: {}", e))
+                    .instrument(span.clone()),
+                );
+            } else {
+                tokio::spawn(
+                    serve::serve(
+                        outbound_listen,
+                        outbound::server::stack(
+                            &outbound,
+                            dst.profiles.clone(),
+                            dst.resolve,
+                            outbound::tcp::connect::stack(
+                                &outbound.proxy.connect,
+                                outbound_addr.port(),
+                                local_identity.clone(),
+                                &outbound_metrics,
+                            ),
+                            outbound_http.clone(),
+                            outbound_metrics,
+                            oc_span_sink.clone(),
+                            drain_rx.clone(),
+                        ),
+                        drain_rx.clone().signal(),
+                    )
+                    .map_err(|e| panic!("outbound failed: {}", e))
+                    .instrument(span.clone()),
+                );
```

Comment:
nit: spent a bit trying to figure out how to reduce repetition here; turns out it's not super easy. but, I think the TCP connect stack could at least be factored out?
```rust
            info!(listen.addr = %outbound_addr, ingress_mode);
            let tcp_connect = outbound::tcp::connect::stack(
                &outbound.proxy.connect,
                outbound_addr.port(),
                local_identity.clone(),
                &outbound_metrics,
            );
            if ingress_mode {
                tokio::spawn(
                    serve::serve(
                        outbound_listen,
                        outbound::ingress::stack(
                            &outbound,
                            dst.profiles.clone(),
                            outbound::tcp::connect::forward(tcp_connect),
                            outbound_http.clone(),
                            &outbound_metrics,
                            oc_span_sink.clone(),
                            drain_rx.clone(),
                        ),
                        drain_rx.clone().signal(),
                    )
                    .map_err(|e| panic!("outbound failed: {}", e))
                    .instrument(span.clone()),
                );
            } else {
                tokio::spawn(
                    serve::serve(
                        outbound_listen,
                        outbound::server::stack(
                            &outbound,
                            dst.profiles.clone(),
                            dst.resolve,
                            tcp_connect,
                            outbound_http.clone(),
                            outbound_metrics,
                            oc_span_sink.clone(),
                            drain_rx.clone(),
                        ),
                        drain_rx.clone().signal(),
                    )
                    .map_err(|e| panic!("outbound failed: {}", e))
                    .instrument(span.clone()),
                );
            }
```